text/0022-gas-spot-market.md

```diff
@@ -0,0 +1,215 @@
+
+(user-side)
+
+1. User submits the deploy to any node as usual.
+
+#### Ordering
+
+(validator-side)
+
+1. Block proposer orders available deploys for inclusion in a protoblock.
+
+    a. Sort by WTP (greatest to least).
+    
+    b. Sort by gas reservation (greatest to least).
+
+#### Balance check & allocation
+
+(validator-side, proposal)
+
+For each channel, starting with WASM-less transfers, the following process is to be followed:
+
+1. Traverse the ordered list of deploys, checking balances.
+    
+    a. Check that originating account purse holds an amount equal to willingness to pay times gas reservation, record the remainder for this originating account purse.
+    
+    b. If sufficient tokens, include the deploy in the protoblock.
+    
+    c. Upon encountering the same originating account purse subsequently, check willingness to pay times gas reservation against remainder, and update the remainder.
+    
+    d. If sufficient tokens in the remainder, include the deploy in the protoblock.
+    
+2. Continue traversing deploy-by-deploy until the gas limit is full.
+
+3. Once gas limit cannot be filled by the next deploy with an eligible balance or remainder, skip deploys until a deploy small enough is encountered to be considered, otherwise continuing as above.
+
+The remainder for each originating account is to carry over from WASM-less transfers to general deploys.
+
+#### Payment code execution
+
+(validator-side, execution)
+
+1. Execute payment code as usual, but recording both originating account purse and purse provided by the payment code.
+
+#### Session code execution
+
+(validator-side, execution)
+
+1. Execute session code as usual, but subject to the available tokens in both the originating account purse and the provided purse.
+
+#### Charges
+
+(validator-side, execution)
+
+1. Charge the provided purse first, charging willingness to pay times gas reservation to the provided purse
+2. If the provided purse has insufficient tokens, charge overflow to the originating account purse according to the rule in 1.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+### Gas requirements estimation
+
+We have already mentioned that, due to the consensus-before-execution model, we cannot charge only for gas used during execution. Charging only for gas used fails to provide any incentive for the users
+to estimate gas requirements and open the proposed system to gaming by setting arbitrarily high gas reservations. This is contrary to the notion high-value transactions be prioritized and could possibly result in consistently underfilled blocks. However, this is a serious UX issue, since we have few tools at present for estimating gas requirements. It is expected that tooling will eventually be developed by third parties, using a testnet as a simulacrum of a live network. Such tooling may be developed from existing tests for simple contracts, but would require considerable work to enable testing of complex assemblages of contracts.
+
```

Comment:
The cargo-casper infrastructure could be leveraged to consume the opcode costs table.   This would not be a huge lift, and then the result of running a suite of tests, it would be possible to return the total gas used in execution.  The cargo-casper infrastructure could also be used to address complex assemblages of contracts as well.